sheets/utils.py

```diff
@@ -40,407 +40,192 @@ def generate_google_client_config():
-            code (str): Coupon code
-            recipient_email (str):
-            event_type (str): The event type (e.g.: "delivered", "failed")
-            event_date (datetime.datetime): The datetime when the email was delivered for the given coupon assignment
-        """
-        assignment_dict = self._assignment_map.get(bulk_assignment_id, {}).get(
-            (code, recipient_email)
-        )
-        if not assignment_dict:
-            return
-        # The "enrolled" status is set by the app when a user redeems a bulk enrollment coupon
-        # and is considered the end of the bulk enrollment flow. It should not be overwritten by any
-        # other status.
-        if assignment_dict["existing_status"] == ASSIGNMENT_SHEET_ENROLLED_STATUS:
-            return
-
-        if (
-            assignment_dict["existing_status"]
-            and assignment_dict["existing_status_date"]
-            and event_type == assignment_dict["existing_status"]
-            and event_date == assignment_dict["existing_status_date"]
-        ):
-            self._assignment_map[bulk_assignment_id][(code, recipient_email)].update(
-                {"new_status": None, "new_status_date": None}
-            )
-        else:
-            self._assignment_map[bulk_assignment_id][(code, recipient_email)].update(
-                {"new_status": event_type, "new_status_date": event_date}
-            )
-
-    def get_new_status_and_date(self, bulk_assignment_id, code, recipient_email):
-        """
-        Returns the new status and status date for the coupon assignment matching the code and email
-        in the given bulk assignment
+class CouponAssignSheetMetadata(
+    SheetMetadata
+):  # pylint: disable=too-many-instance-attributes
+    """Metadata for a coupon assignment spreadsheet"""
 
-        Args:
-            bulk_assignment_id (int):
-            code (str): Coupon code
-            recipient_email (str):
+    def __init__(self):
+        self.sheet_type = SHEET_TYPE_COUPON_ASSIGN
+        self.sheet_name = "Coupon Assignment sheet"
+        self.first_data_row = GOOGLE_SHEET_FIRST_ROW + 1
+        self.column_headers = [
+            "Coupon Code",
+            "Email (Assignee)",
+            "Status",
+            "Status Date",
+        ]
+        self.non_input_column_indices = {0, 2, 3}
+        self.num_columns = len(self.column_headers)
 
-        Returns:
-            (str or None, datetime.datetime or None): The new status paired with the date that
-                the status was logged in Mailgun (or (None, None) if the given assignment does not
-                have a new status.
-        """
-        message_data = self._assignment_map[bulk_assignment_id].get(
-            (code, recipient_email)
-        )
-        return (
-            (message_data["new_status"], message_data["new_status_date"])
-            if message_data
-            else (None, None)
+        self.LAST_COL_LETTER = get_column_letter(self.num_columns - 1)
```

Comment:
Pygsheets documentation says that it wants an uppercase value to indicate the column in a spreadsheet cell reference. Maybe it works with lowercase too, but I don't see any benefit from attempting it anyway. For the status column header values, those are uppercase-first because those values are used directly to fill in the header row in a spreadsheet generated by this app. The actual value of `STATUS_COL` is in integer index